### PR TITLE
feat(auth): implement addMachineIfAbsent and pass unit test

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

* Remove the statemachines from the manager contructor, and add an `addMachineIfAbsent()` to support adding statemachine dynamically. We can add the state machines to the class contructor later of there's an use case
* Add more unit tests and make sure the unit test passes. There will be followup PR to unblock the CI
* Fix formatting.

#### Description of how you validated changes
```bash
cd packages/auth && yarn jest __tests__/state-machine-*.ts
yarn run v1.22.19
$ .../amplify-js/node_modules/.bin/jest __tests__/state-machine-manager-test.ts __tests__/state-machine-single-test.ts
 PASS  __tests__/state-machine-single-test.ts
 PASS  __tests__/state-machine-manager-test.ts (10.972s)
````

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/stateMachine/stateMachineManager.ts [hunk 1]
```diff
@@ -32,41 +49,80 @@ type MachinePromiseContext = [
...
+	): Promise<CurrentStateAndContext<MachineContext, string>> {
+		return await this._enqueueEvent({
+			name: CURRENT_STATES_EVENT_SYMBOL,
+			payload: {},
+			toMachine: machineName,
+		});
+	}
+
+	/**
+	 * Add a new state machine to the manager. If there are events being processed,
```
Comment (haverchuck) [inline packages/auth/src/stateMachine/stateMachineManager.ts:85]: 👍

**Review by haverchuck — APPROVED** [review]

File: packages/auth/src/stateMachine/stateMachineManager.ts [hunk 2]
```diff
@@ -9,21 +9,38 @@ import {
 	MachineManagerEvent,
 } from './types';
 import { Machine } from './machine';
+import { Logger } from '@aws-amplify/core';
```
Comment (jimblanc) [inline packages/auth/src/stateMachine/stateMachineManager.ts:12]: Nit but would prefer to put non-relative imports first

File: packages/auth/src/stateMachine/stateMachineManager.ts [hunk 3]
```diff
@@ -9,21 +9,38 @@ import {
...
 
 type MachineState = CurrentStateAndContext<MachineContext, string>;
+// General type for any machine.
+type MachineType = Machine<any, any, any>;
```
Comment (jimblanc) [inline packages/auth/src/stateMachine/stateMachineManager.ts:16]: Good for now but any way to more strongly type this?

File: packages/auth/src/stateMachine/stateMachineManager.ts [hunk 4]
```diff
@@ -9,21 +9,38 @@ import {
...
 
 const CURRENT_STATES_EVENT_SYMBOL = Symbol('CURRENT_STATES_EVENT_PAYLOAD');
+interface CurrentStateEventType extends Omit<MachineManagerEvent, 'name'> {
+	name: typeof CURRENT_STATES_EVENT_SYMBOL;
+}
+
+const ADD_MACHINE_EVENT_SYMBOL = Symbol('ADD_MACHINE_EVENT_PAYLOAD');
+interface AddMachineEventType extends Omit<MachineManagerEvent, 'name'> {
+	name: typeof ADD_MACHINE_EVENT_SYMBOL;
+	payload: { machine: Machine<any, any, any> };
```
Comment (jimblanc) [inline packages/auth/src/stateMachine/stateMachineManager.ts:26]: This should be `MachineType` I think

File: packages/auth/src/stateMachine/stateMachineManager.ts [hunk 5]
```diff
@@ -116,12 +178,13 @@ export class MachineManager<MachineTypes extends Machine<any, any, any>> {
 		}
 		const machine = this._machines[event.toMachine];
 		if (!machine) {
-			// TODO: remove this error but put this message into logger.
-			throw new Error(
+			this.logger.debug(
 				`Cannot route event to machine ${
 					event.toMachine
 				}. Event ${JSON.stringify(event)}`
```
Comment (jimblanc) [inline packages/auth/src/stateMachine/stateMachineManager.ts:184]: Any risk of leaking info from the serialized event? Granted it is a debug log statement, but would the event type be enough?

**Review by jimblanc — APPROVED** [review]

File: packages/auth/src/stateMachine/stateMachineManager.ts (hunk 5 above)
Comment (haverchuck) [reply packages/auth/src/stateMachine/stateMachineManager.ts:184]: Good call - the event might contain sensitive data like username/password.

Comment (AllanZhengYP) [reply packages/auth/src/stateMachine/stateMachineManager.ts:184]: This is a good callout. I will change to only print the name and event id only

**Review by jimblanc — APPROVED** [review]

**Pull request merged**